File: go/pkg/proxy/socks5.go
// Package proxy implements SOCKS5 and HTTP CONNECT proxy servers.
//
// The proxy server accepts client connections (SOCKS5 or HTTP CONNECT),
// parses target addresses, and dials remote targets through a pluggable
// DialFunc. For zgrnet integration, the DialFunc opens KCP streams via
// the UDP high-level API with proto=TCP_PROXY(69).
package proxy

import (
	"bufio"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"net"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/vibing/zgrnet/pkg/noise"
)

// SOCKS5 protocol constants.
const (
	Version5 = 0x05

	// Authentication methods.
	AuthNone     = 0x00
	AuthNoAccept = 0xFF

	// Commands.
	CmdConnect      = 0x01
	CmdBind         = 0x02
	CmdUDPAssociate = 0x03

	// Reply codes.
	RepSuccess          = 0x00
	RepGeneralFailure   = 0x01
	RepNotAllowed       = 0x02
	RepNetworkUnreach   = 0x03
	RepHostUnreach      = 0x04
	RepConnRefused      = 0x05
	RepTTLExpired       = 0x06
	RepCmdNotSupported  = 0x07
	RepAddrNotSupported = 0x08
)

// Errors.
var (
	ErrUnsupportedVersion = errors.New("proxy: unsupported SOCKS version")
	ErrNoAcceptableAuth   = errors.New("proxy: no acceptable auth method")
	ErrUnsupportedCommand = errors.New("proxy: unsupported command")
	ErrInvalidAddress     = errors.New("proxy: invalid address")
)

// DialFunc opens a connection to the target address.
// The returned io.ReadWriteCloser must have blocking Read semantics.
type DialFunc func(addr *noise.Address) (io.ReadWriteCloser, error)

// UDPRelay handles UDP data exchange through the tunnel.
type UDPRelay interface {
	// WriteTo sends UDP data to the target address through the tunnel.
	WriteTo(addr *noise.Address, data []byte) error
	// ReadFrom reads a UDP response from the tunnel.
	// Returns source address, number of data bytes copied to buf, and error.
	ReadFrom(buf []byte) (addr *noise.Address, n int, err error)
	// Close closes the relay.
	Close() error
}

// NewUDPRelayFunc creates a new UDPRelay for a UDP association.
// Each UDP ASSOCIATE connection gets its own relay instance.
type NewUDPRelayFunc func() (UDPRelay, error)

// ProxyStats holds proxy connection statistics.
type ProxyStats struct {
	TotalConnections uint64 `json:"total_connections"`
	ActiveConnections int64 `json:"active_connections"`
	BytesSent        uint64 `json:"bytes_sent"`
	BytesReceived    uint64 `json:"bytes_received"`
	Errors           uint64 `json:"errors"`
}

// Server is a SOCKS5 proxy server.
type Server struct {
	listenAddr string
	dial       DialFunc        // For TCP CONNECT
	newRelay   NewUDPRelayFunc // For UDP ASSOCIATE (nil = not supported)
	policy     Policy          // nil = allow all

	mu       sync.Mutex
	listener net.Listener

	closed atomic.Bool
	wg     sync.WaitGroup

	// Atomic stats counters
	totalConns  atomic.Uint64
	activeConns atomic.Int64
	bytesSent   atomic.Uint64
	bytesRecv   atomic.Uint64
	errors      atomic.Uint64
}

// GetStats returns a snapshot of proxy connection statistics.
func (s *Server) GetStats() ProxyStats {
	return ProxyStats{
		TotalConnections:  s.totalConns.Load(),
		ActiveConnections: s.activeConns.Load(),
		BytesSent:         s.bytesSent.Load(),
		BytesReceived:     s.bytesRecv.Load(),
		Errors:            s.errors.Load(),
	}
}

// SetPolicy sets the policy for target address validation.
// If nil (default), all addresses are allowed.
func (s *Server) SetPolicy(p Policy) {
	s.policy = p
}

// NewServer creates a new proxy server.
// listenAddr is the TCP address to listen on (e.g. "127.0.0.1:1080").
// dial is the function used to connect to target addresses.
func NewServer(listenAddr string, dial DialFunc) *Server {
	return &Server{
		listenAddr: listenAddr,
		dial:       dial,
	}
}

// ListenAndServe binds to the configured address and starts serving.
func (s *Server) ListenAndServe() error {
	ln, err := net.Listen("tcp", s.listenAddr)
	if err != nil {
		return err
	}
	return s.Serve(ln)
}

// Serve accepts connections on the given listener.
func (s *Server) Serve(ln net.Listener) error {
	s.mu.Lock()
	s.listener = ln
	s.mu.Unlock()

	for {
		conn, err := ln.Accept()
		if err != nil {
			if s.closed.Load() {
				return nil
			}
			continue
		}

		s.wg.Add(1)
		s.totalConns.Add(1)
		s.activeConns.Add(1)
		go func() {
			defer s.wg.Done()
			defer s.activeConns.Add(-1)
			s.handleConn(conn)
		}()
	}
}

// Addr returns the listener address. Returns nil if not listening.
func (s *Server) Addr() net.Addr {
	s.mu.Lock()
	ln := s.listener
	s.mu.Unlock()
	if ln == nil {
		return nil
	}
	return ln.Addr()
}

// Close stops the server and waits for active connections to finish.
func (s *Server) Close() error {
	if s.closed.Swap(true) {
		return nil
	}
	s.mu.Lock()
	ln := s.listener
	s.mu.Unlock()
	var err error
	if ln != nil {
		err = ln.Close()
	}
	s.wg.Wait()
	return err
}

// handleConn processes a single client connection.
// Detects SOCKS5 (0x05) vs HTTP CONNECT ('C') by first byte.
func (s *Server) handleConn(conn net.Conn) {
	defer conn.Close()

	// Set a deadline for the handshake phase
	conn.SetDeadline(time.Now().Add(10 * time.Second))

	// Read first byte to detect protocol
	var first [1]byte
	if _, err := io.ReadFull(conn, first[:]); err != nil {
		return
	}

	switch first[0] {
	case Version5:
		s.handleSOCKS5(conn)
	case 'C':
		s.handleHTTPConnect(conn)
	default:
		// Unknown protocol
		return
	}
}

// handleSOCKS5 processes a SOCKS5 connection (version byte already consumed).
func (s *Server) handleSOCKS5(conn net.Conn) {
	// === Auth negotiation ===
	// Read NMETHODS
	var nMethods [1]byte
	if _, err := io.ReadFull(conn, nMethods[:]); err != nil {
		return
	}

	// Read method list
	methods := make([]byte, nMethods[0])
	if _, err := io.ReadFull(conn, methods); err != nil {
		return
	}

	// Check for NO AUTH (0x00)
	hasNoAuth := false
	for _, m := range methods {
		if m == AuthNone {
			hasNoAuth = true
			break
		}
	}

	if !hasNoAuth {
		conn.Write([]byte{Version5, AuthNoAccept})
		return
	}

	// Accept NO AUTH
	if _, err := conn.Write([]byte{Version5, AuthNone}); err != nil {
		return
	}

	// === Request ===
	// VER(1) + CMD(1) + RSV(1) + ATYP(1) = 4 bytes
	var header [4]byte
	if _, err := io.ReadFull(conn, header[:]); err != nil {
		return
	}

	if header[0] != Version5 {
		return
	}

	cmd := header[1]
	atyp := header[3]

	// Parse address
	addr, err := ReadAddress(conn, atyp)
	if err != nil {
		sendReply(conn, RepAddrNotSupported, nil)
		return
	}

	// Clear deadline for the relay phase
	conn.SetDeadline(time.Time{})

	switch cmd {
	case CmdConnect:
		s.handleConnect(conn, addr)
	case CmdUDPAssociate:
		s.handleUDPAssociate(conn, addr)
	default:
		sendReply(conn, RepCmdNotSupported, nil)
	}
}

// handleConnect handles the SOCKS5 CONNECT command.
func (s *Server) handleConnect(conn net.Conn, addr *noise.Address) {
	if !checkPolicy(s.policy, addr) {
		sendReply(conn, RepNotAllowed, nil)
		return
	}

	// Dial the target through the tunnel
	remote, err := s.dial(addr)
	if err != nil {
		sendReply(conn, RepGeneralFailure, nil)
		return
	}
	defer remote.Close()

	// Send success reply with the target address as bound address
	sendReply(conn, RepSuccess, addr)

	// Relay data bidirectionally
	Relay(conn, remote)
}

// handleUDPAssociate handles the SOCKS5 UDP ASSOCIATE command.
//
// The server binds a local UDP port, returns it to the client, then relays
// UDP datagrams between the client (SOCKS5 UDP format) and the tunnel
// (UDPRelay). The TCP connection serves as a control channel — when it
// closes, the association ends.
func (s *Server) handleUDPAssociate(conn net.Conn, addr *noise.Address) {
	if s.newRelay == nil {
		sendReply(conn, RepCmdNotSupported, nil)
		return
	}

	// Create a relay for this association
	relay, err := s.newRelay()
	if err != nil {
		sendReply(conn, RepGeneralFailure, nil)
		return
	}

	// Determine bind IP from TCP listener (matches the SOCKS5 listen address)
	var bindIP net.IP
	s.mu.Lock()
	if s.listener != nil {
		if tcpAddr, ok := s.listener.Addr().(*net.TCPAddr); ok {
			bindIP = tcpAddr.IP
		}
	}
	s.mu.Unlock()
	if bindIP == nil {
		bindIP = net.IPv4(127, 0, 0, 1)
	}

	// Bind a local UDP socket for the client
	udpConn, err := net.ListenUDP("udp", &net.UDPAddr{IP: bindIP})
	if err != nil {
		relay.Close()
		sendReply(conn, RepGeneralFailure, nil)
		return
	}

	// Send bound address to client
	boundAddr := udpConn.LocalAddr().(*net.UDPAddr)
	replyAddr := &noise.Address{
		Type: noise.AddressTypeIPv4,
		Host: boundAddr.IP.String(),
		Port: uint16(boundAddr.Port),
	}
	sendReply(conn, RepSuccess, replyAddr)

	// Track the client's UDP address (set on first packet received)
	var clientAddr atomic.Pointer[net.UDPAddr]

	var wg sync.WaitGroup

	// Outbound: client UDP → tunnel
	wg.Add(1)
	go func() {
		defer wg.Done()
		buf := make([]byte, 65535)
		for {
			n, from, err := udpConn.ReadFromUDP(buf)
			if err != nil {
				return
			}
			clientAddr.Store(from)

			targetAddr, data, err := ParseSOCKS5UDP(buf[:n])
			if err != nil {
				continue
			}
			relay.WriteTo(targetAddr, data)
		}
	}()

	// Inbound: tunnel → client UDP
	wg.Add(1)
	go func() {
		defer wg.Done()
		buf := make([]byte, 65535)
		for {
			srcAddr, n, err := relay.ReadFrom(buf)
			if err != nil {
				return
			}

			ca := clientAddr.Load()
			if ca == nil {
				continue // No client connected yet
			}

			pkt := BuildSOCKS5UDP(srcAddr, buf[:n])
			if pkt != nil {
				udpConn.WriteToUDP(pkt, ca)
			}
		}
	}()

	// Wait for TCP control connection to close (signals end of association)
	io.Copy(io.Discard, conn)

	// Cleanup: close sockets to unblock goroutines, then wait
	udpConn.Close()
	relay.Close()
	wg.Wait()
}

// handleHTTPConnect handles an HTTP CONNECT request.
//
// Format: "CONNECT host:port HTTP/1.x\r\n" + headers + "\r\n"
// The first byte 'C' has already been consumed by handleConn.
func (s *Server) handleHTTPConnect(conn net.Conn) {
	// Use a small bufio.Reader for header parsing.
	// The 4KB default buffer limits how much data can accumulate
	// before a newline, preventing unbounded memory growth from
	// malicious clients sending very long header lines.
	reader := bufio.NewReaderSize(conn, 4096)

	// Read the rest of the first line (we already consumed 'C')
	restOfLine, err := reader.ReadString('\n')
	if err != nil {
		return
	}
	firstLine := "C" + restOfLine

	// Parse: "CONNECT host:port HTTP/1.x\r\n"
	parts := strings.SplitN(strings.TrimSpace(firstLine), " ", 3)
	if len(parts) < 2 || parts[0] != "CONNECT" {
		conn.Write([]byte("HTTP/1.1 400 Bad Request\r\n\r\n"))
		return
	}

	target := parts[1] // "host:port"

	// Read remaining headers until empty line
	for {
		line, err := reader.ReadString('\n')
		if err != nil {
			return
		}
		if line == "\r\n" || line == "\n" {
			break
		}
	}

	// Parse target address
	host, portStr, err := net.SplitHostPort(target)
	if err != nil {
		// Try as host-only (default port 443 for CONNECT)
		host = target
		portStr = "443"
	}
	port, err := strconv.Atoi(portStr)
	if err != nil || port < 1 || port > 65535 {
		conn.Write([]byte("HTTP/1.1 400 Bad Request\r\n\r\n"))
		return
	}

	// Determine address type
	addr := &noise.Address{Port: uint16(port)}
	if ip := net.ParseIP(host); ip != nil {
		if ip.To4() != nil {
			addr.Type = noise.AddressTypeIPv4
		} else {
			addr.Type = noise.AddressTypeIPv6
		}
		addr.Host = ip.String()
	} else {
		addr.Type = noise.AddressTypeDomain
		addr.Host = host
	}

	// Clear deadline for relay phase
	conn.SetDeadline(time.Time{})

	// Policy check
	if !checkPolicy(s.policy, addr) {
		conn.Write([]byte("HTTP/1.1 403 Forbidden\r\n\r\n"))
		return
	}

	// Dial the target
	remote, err := s.dial(addr)
	if err != nil {
		conn.Write([]byte("HTTP/1.1 502 Bad Gateway\r\n\r\n"))
		return
	}
	defer remote.Close()

	// Send success response
	conn.Write([]byte("HTTP/1.1 200 Connection Established\r\n\r\n"))

	// Relay using buffered reader (may have data buffered beyond headers)
	bc := &bufferedConn{reader: reader, conn: conn}
	Relay(bc, remote)
}

// bufferedConn wraps a net.Conn with a bufio.Reader for reads.
// This ensures any data buffered during HTTP header parsing is not lost.
type bufferedConn struct {
	reader *bufio.Reader
	conn   net.Conn
}

func (c *bufferedConn) Read(p []byte) (int, error)  { return c.reader.Read(p) }
func (c *bufferedConn) Write(p []byte) (int, error) { return c.conn.Write(p) }
func (c *bufferedConn) Close() error                { return c.conn.Close() }

// CloseWrite supports half-close for the relay.
func (c *bufferedConn) CloseWrite() error {
	if tc, ok := c.conn.(*net.TCPConn); ok {
		return tc.CloseWrite()
	}
	return nil
}

// sendReply sends a SOCKS5 reply to the client.
// Format: VER(1) + REP(1) + RSV(1) + ATYP(1) + BND.ADDR(var) + BND.PORT(2)
func sendReply(conn net.Conn, rep byte, addr *noise.Address) {
	if addr != nil {
		if encoded := addr.Encode(); encoded != nil {
			reply := make([]byte, 3+len(encoded))
			reply[0] = Version5
			reply[1] = rep
			reply[2] = 0x00 // RSV
			copy(reply[3:], encoded)
			conn.Write(reply)
			return
		}
	}

	// Default bound address: 0.0.0.0:0
	conn.Write([]byte{Version5, rep, 0x00, noise.AddressTypeIPv4, 0, 0, 0, 0, 0, 0})
}

// ReadAddress reads a SOCKS5 address from a reader.
// atyp is the address type byte already read from the stream.
func ReadAddress(r io.Reader, atyp byte) (*noise.Address, error) {
	switch atyp {
	case noise.AddressTypeIPv4:
		var buf [6]byte // 4 bytes IP + 2 bytes port
		if _, err := io.ReadFull(r, buf[:]); err != nil {
			return nil, err
		}
		return &noise.Address{
			Type: noise.AddressTypeIPv4,
			Host: net.IP(buf[:4]).String(),
			Port: binary.BigEndian.Uint16(buf[4:6]),
		}, nil

	case noise.AddressTypeDomain:
		var lenBuf [1]byte
		if _, err := io.ReadFull(r, lenBuf[:]); err != nil {
			return nil, err
		}
		domainLen := int(lenBuf[0])
		if domainLen == 0 {
			return nil, ErrInvalidAddress
		}
		buf := make([]byte, domainLen+2) // domain + 2 bytes port
		if _, err := io.ReadFull(r, buf); err != nil {
			return nil, err
		}
		return &noise.Address{
			Type: noise.AddressTypeDomain,
			Host: string(buf[:domainLen]),
			Port: binary.BigEndian.Uint16(buf[domainLen:]),
		}, nil

	case noise.AddressTypeIPv6:
		var buf [18]byte // 16 bytes IP + 2 bytes port
		if _, err := io.ReadFull(r, buf[:]); err != nil {
			return nil, err
		}
		return &noise.Address{
			Type: noise.AddressTypeIPv6,
			Host: net.IP(buf[:16]).String(),
			Port: binary.BigEndian.Uint16(buf[16:18]),
		}, nil

	default:
		return nil, fmt.Errorf("proxy: unsupported address type 0x%02x", atyp)
	}
}

// Relay copies data bidirectionally between two ReadWriteClosers.
// Returns when either direction encounters an error or EOF.
// Supports half-close via CloseWrite if the underlying type supports it.
func Relay(a, b io.ReadWriteCloser) {
	var wg sync.WaitGroup
	wg.Add(2)

	// a → b
	go func() {
		defer wg.Done()
		io.Copy(b, a)
		if tc, ok := b.(interface{ CloseWrite() error }); ok {
			tc.CloseWrite()
		}
	}()

	// b → a
	go func() {
		defer wg.Done()
		io.Copy(a, b)
		if tc, ok := a.(interface{ CloseWrite() error }); ok {
			tc.CloseWrite()
		}
	}()

	wg.Wait()
}

// ParseSOCKS5UDP parses a SOCKS5 UDP datagram.
// Format: RSV(2) + FRAG(1) + ATYP(1) + DST.ADDR(var) + DST.PORT(2) + DATA(var)
// Returns the target address and the data payload.
func ParseSOCKS5UDP(data []byte) (addr *noise.Address, payload []byte, err error) {
	if len(data) < 4 {
		return nil, nil, errors.New("proxy: UDP datagram too short")
	}
	if data[0] != 0 || data[1] != 0 {
		return nil, nil, errors.New("proxy: invalid RSV in UDP datagram")
	}
	if data[2] != 0 {
		return nil, nil, errors.New("proxy: fragmented UDP not supported")
	}
	// Decode address starting at byte 3 (atyp + addr + port)
	addr, consumed, err := noise.DecodeAddress(data[3:])
	if err != nil {
		return nil, nil, err
	}
	payload = data[3+consumed:]
	return addr, payload, nil
}

// BuildSOCKS5UDP builds a SOCKS5 UDP datagram.
// Format: RSV(2) + FRAG(1) + ATYP(1) + SRC.ADDR(var) + SRC.PORT(2) + DATA(var)
func BuildSOCKS5UDP(addr *noise.Address, data []byte) []byte {
	encoded := addr.Encode()
	if encoded == nil {
		return nil
	}
	pkt := make([]byte, 3+len(encoded)+len(data))
	// RSV(2) + FRAG(1) = 0x00 0x00 0x00
	copy(pkt[3:], encoded)
	copy(pkt[3+len(encoded):], data)
	return pkt
}

File: zig/src/proxy/socks5.zig
//! SOCKS5 and HTTP CONNECT proxy protocol implementation.
//!
//! Provides SOCKS5 address parsing, reply building, and UDP datagram helpers.
//! The server accept loop is left to the caller; this module focuses on
//! protocol parsing and formatting that is independent of the I/O backend.

const std = @import("std");
const noise_address = @import("../noise/address.zig");
const Address = noise_address.Address;
const AddressError = noise_address.AddressError;
const ATYP_IPV4 = noise_address.ATYP_IPV4;
const ATYP_DOMAIN = noise_address.ATYP_DOMAIN;
const ATYP_IPV6 = noise_address.ATYP_IPV6;

// SOCKS5 protocol constants
pub const VERSION5: u8 = 0x05;
pub const AUTH_NONE: u8 = 0x00;
pub const AUTH_NO_ACCEPT: u8 = 0xFF;

pub const CMD_CONNECT: u8 = 0x01;
pub const CMD_BIND: u8 = 0x02;
pub const CMD_UDP_ASSOCIATE: u8 = 0x03;

pub const REP_SUCCESS: u8 = 0x00;
pub const REP_GENERAL_FAILURE: u8 = 0x01;
pub const REP_NOT_ALLOWED: u8 = 0x02;
pub const REP_NETWORK_UNREACH: u8 = 0x03;
pub const REP_HOST_UNREACH: u8 = 0x04;
pub const REP_CONN_REFUSED: u8 = 0x05;
pub const REP_TTL_EXPIRED: u8 = 0x06;
pub const REP_CMD_NOT_SUPPORTED: u8 = 0x07;
pub const REP_ADDR_NOT_SUPPORTED: u8 = 0x08;

pub const Error = error{
    InvalidProtocol,
    InvalidAuth,
    UnsupportedCommand,
    InvalidAddress,
    FragmentNotSupported,
    DatagramTooShort,
    InvalidRSV,
    ConnectionRefused,
    EndOfStream,
    BufferTooSmall,
};

/// SOCKS5 request parsed from the wire.
pub const Request = struct {
    cmd: u8,
    addr: Address,
};

/// Read SOCKS5 auth negotiation and accept NO AUTH.
/// Returns error if no acceptable method is found.
pub fn negotiateAuth(reader: anytype, writer: anytype) !void {
    // Read NMETHODS
    var n_methods_buf: [1]u8 = undefined;
    _ = try reader.readAll(&n_methods_buf);
    const n_methods = n_methods_buf[0];

    // Read methods
    var methods_buf: [256]u8 = undefined;
    const methods = methods_buf[0..n_methods];
    _ = try reader.readAll(methods);

    // Check for NO AUTH
    var has_no_auth = false;
    for (methods) |m| {
        if (m == AUTH_NONE) {
            has_no_auth = true;
            break;
        }
    }

    if (!has_no_auth) {
        _ = try writer.write(&[_]u8{ VERSION5, AUTH_NO_ACCEPT });
        return Error.InvalidAuth;
    }

    _ = try writer.write(&[_]u8{ VERSION5, AUTH_NONE });
}

/// Read a SOCKS5 request (after auth negotiation).
/// Returns the command and target address.
pub fn readRequest(reader: anytype, host_buf: []u8) !Request {
    var header: [4]u8 = undefined;
    _ = try reader.readAll(&header);

    if (header[0] != VERSION5) {
        return Error.InvalidProtocol;
    }

    const cmd = header[1];
    const atyp = header[3];

    const addr = try readAddress(reader, atyp, host_buf);
    return .{ .cmd = cmd, .addr = addr };
}

/// Read a SOCKS5 address from a stream.
pub fn readAddress(reader: anytype, atyp: u8, host_buf: []u8) !Address {
    switch (atyp) {
        ATYP_IPV4 => {
            var buf: [6]u8 = undefined;
            _ = try reader.readAll(&buf);
            const ip_str = std.fmt.bufPrint(host_buf, "{}.{}.{}.{}", .{
                buf[0], buf[1], buf[2], buf[3],
            }) catch return Error.BufferTooSmall;
            const port = std.mem.readInt(u16, buf[4..6], .big);
            return .{ .atyp = ATYP_IPV4, .host = ip_str, .port = port };
        },
        ATYP_DOMAIN => {
            var len_buf: [1]u8 = undefined;
            _ = try reader.readAll(&len_buf);
            const domain_len: usize = len_buf[0];
            if (domain_len == 0) return Error.InvalidAddress;

            var buf: [257]u8 = undefined; // max domain 255 + 2 port
            _ = try reader.readAll(buf[0 .. domain_len + 2]);
            const port = std.mem.readInt(u16, buf[domain_len..][0..2], .big);
            // Copy domain to host_buf
            if (host_buf.len < domain_len) return Error.BufferTooSmall;
            @memcpy(host_buf[0..domain_len], buf[0..domain_len]);
            return .{ .atyp = ATYP_DOMAIN, .host = host_buf[0..domain_len], .port = port };
        },
        ATYP_IPV6 => {
            var buf: [18]u8 = undefined;
            _ = try reader.readAll(&buf);
            const octets: [16]u8 = buf[0..16].*;
            const ip6 = std.net.Ip6Address.init(octets, 0, 0, 0);
            // Format IPv6 address
            var tmp: [64]u8 = undefined;
            const full = std.fmt.bufPrint(&tmp, "{f}", .{ip6}) catch return Error.BufferTooSmall;
            // Strip brackets: "[addr]:0" → "addr"
            if (full.len < 4 or full[0] != '[') return Error.InvalidAddress;
            const close = std.mem.indexOfScalar(u8, full, ']') orelse return Error.InvalidAddress;
            const bare = full[1..close];
            if (host_buf.len < bare.len) return Error.BufferTooSmall;
            @memcpy(host_buf[0..bare.len], bare);
            const port = std.mem.readInt(u16, buf[16..18], .big);
            return .{ .atyp = ATYP_IPV6, .host = host_buf[0..bare.len], .port = port };
        },
        else => return Error.InvalidAddress,
    }
}

/// Build a SOCKS5 reply into a buffer.
/// Returns the slice written.
pub fn buildReply(buf: []u8, rep: u8, addr: ?*const Address) ![]u8 {
    if (addr) |a| {
        const encoded_size = a.encodedSize() catch 7;
        if (buf.len < 3 + encoded_size) return Error.BufferTooSmall;
        buf[0] = VERSION5;
        buf[1] = rep;
        buf[2] = 0x00;
        _ = a.encode(buf[3..]) catch {
            // Fallback to 0.0.0.0:0
            buf[3] = ATYP_IPV4;
            @memset(buf[4..10], 0);
            return buf[0..10];
        };
        return buf[0 .. 3 + encoded_size];
    } else {
        // Default: 0.0.0.0:0
        if (buf.len < 10) return Error.BufferTooSmall;
        buf[0] = VERSION5;
        buf[1] = rep;
        buf[2] = 0x00;
        buf[3] = ATYP_IPV4;
        @memset(buf[4..10], 0);
        return buf[0..10];
    }
}

/// Parse a SOCKS5 UDP datagram.
/// Format: RSV(2) + FRAG(1) + ATYP(1) + ADDR(var) + PORT(2) + DATA(var)
/// Returns the target address and payload slice.
pub fn parseSOCKS5UDP(data: []const u8, host_buf: []u8) !struct { addr: Address, payload: []const u8 } {
    if (data.len < 4) return Error.DatagramTooShort;
    if (data[0] != 0 or data[1] != 0) return Error.InvalidRSV;
    if (data[2] != 0) return Error.FragmentNotSupported;

    const result = try Address.decode(data[3..], host_buf);
    const payload = data[3 + result.consumed ..];
    return .{ .addr = result.addr, .payload = payload };
}

/// Build a SOCKS5 UDP datagram.
/// Format: RSV(2) + FRAG(1) + encoded_addr + DATA
/// Returns the slice written into buf.
pub fn buildSOCKS5UDP(buf: []u8, addr: *const Address, data: []const u8) ![]u8 {
    const addr_size = try addr.encodedSize();
    const total = 3 + addr_size + data.len;
    if (buf.len < total) return Error.BufferTooSmall;

    buf[0] = 0x00; // RSV
    buf[1] = 0x00; // RSV
    buf[2] = 0x00; // FRAG
    _ = try addr.encode(buf[3..]);
    @memcpy(buf[3 + addr_size .. 3 + addr_size + data.len], data);
    return buf[0..total];
}

// =============================================================================
// Tests
// =============================================================================

test "negotiateAuth NO AUTH accepted" {
    // Simulate client sending: NMETHODS=1, METHODS=[0x00]
    const input = [_]u8{ 0x01, 0x00 };
    var reader = std.io.fixedBufferStream(&input);
    var output_buf: [64]u8 = undefined;
    var writer = std.io.fixedBufferStream(&output_buf);

    try negotiateAuth(reader.reader(), writer.writer());

    const written = writer.getWritten();
    try std.testing.expectEqual(@as(usize, 2), written.len);
    try std.testing.expectEqual(VERSION5, written[0]);
    try std.testing.expectEqual(AUTH_NONE, written[1]);
}

test "negotiateAuth reject" {
    // Only offer USER/PASS (0x02)
    const input = [_]u8{ 0x01, 0x02 };
    var reader = std.io.fixedBufferStream(&input);
    var output_buf: [64]u8 = undefined;
    var writer = std.io.fixedBufferStream(&output_buf);

    const result = negotiateAuth(reader.reader(), writer.writer());
    try std.testing.expectError(Error.InvalidAuth, result);

    const written = writer.getWritten();
    try std.testing.expectEqual(VERSION5, written[0]);
    try std.testing.expectEqual(AUTH_NO_ACCEPT, written[1]);
}

test "readRequest CONNECT IPv4" {
    // VER=5, CMD=CONNECT, RSV=0, ATYP=IPv4, 10.0.0.1:8080
    const input = [_]u8{
        0x05, CMD_CONNECT, 0x00, ATYP_IPV4,
        10,   0,           0,    1,
        0x1F, 0x90, // port 8080
    };
    var reader = std.io.fixedBufferStream(&input);
    var host_buf: [64]u8 = undefined;

    const req = try readRequest(reader.reader(), &host_buf);
    try std.testing.expectEqual(CMD_CONNECT, req.cmd);
    try std.testing.expectEqualStrings("10.0.0.1", req.addr.host);
    try std.testing.expectEqual(@as(u16, 8080), req.addr.port);
}

test "readRequest CONNECT domain" {
    const domain = "example.com";
    const input = [_]u8{
        0x05,             CMD_CONNECT, 0x00, ATYP_DOMAIN,
        @intCast(domain.len),
    } ++ domain.* ++ [_]u8{ 0x01, 0xBB }; // port 443

    var reader = std.io.fixedBufferStream(&input);
    var host_buf: [64]u8 = undefined;

    const req = try readRequest(reader.reader(), &host_buf);
    try std.testing.expectEqual(CMD_CONNECT, req.cmd);
    try std.testing.expectEqualStrings("example.com", req.addr.host);
    try std.testing.expectEqual(@as(u16, 443), req.addr.port);
}

test "buildReply success with address" {
    const addr = Address.ipv4("10.0.0.1", 80);
    var buf: [64]u8 = undefined;
    const reply = try buildReply(&buf, REP_SUCCESS, &addr);
    try std.testing.expectEqual(VERSION5, reply[0]);
    try std.testing.expectEqual(REP_SUCCESS, reply[1]);
    try std.testing.expectEqual(@as(u8, 0x00), reply[2]);
    try std.testing.expectEqual(ATYP_IPV4, reply[3]);
}

test "buildReply failure without address" {
    var buf: [64]u8 = undefined;
    const reply = try buildReply(&buf, REP_GENERAL_FAILURE, null);
    try std.testing.expectEqual(@as(usize, 10), reply.len);
    try std.testing.expectEqual(VERSION5, reply[0]);
    try std.testing.expectEqual(REP_GENERAL_FAILURE, reply[1]);
    try std.testing.expectEqual(ATYP_IPV4, reply[3]);
}

test "parseSOCKS5UDP roundtrip" {
    const addr = Address.ipv4("10.0.0.1", 53);
    const data = "dns query";
    var buf: [128]u8 = undefined;
    const pkt = try buildSOCKS5UDP(&buf, &addr, data);

    var host_buf: [64]u8 = undefined;
    const result = try parseSOCKS5UDP(pkt, &host_buf);
    try std.testing.expectEqualStrings("10.0.0.1", result.addr.host);
    try std.testing.expectEqual(@as(u16, 53), result.addr.port);
    try std.testing.expectEqualStrings("dns query", result.payload);
}

test "parseSOCKS5UDP domain roundtrip" {
    const addr = Address.domain("example.com", 443);
    const data = "test payload";
    var buf: [128]u8 = undefined;
    const pkt = try buildSOCKS5UDP(&buf, &addr, data);

    var host_buf: [64]u8 = undefined;
    const result = try parseSOCKS5UDP(pkt, &host_buf);
    try std.testing.expectEqualStrings("example.com", result.addr.host);
    try std.testing.expectEqual(@as(u16, 443), result.addr.port);
    try std.testing.expectEqualStrings("test payload", result.payload);
}

test "parseSOCKS5UDP too short" {
    var host_buf: [64]u8 = undefined;
    try std.testing.expectError(Error.DatagramTooShort, parseSOCKS5UDP(&[_]u8{ 0, 0 }, &host_buf));
}

test "parseSOCKS5UDP invalid RSV" {
    var host_buf: [64]u8 = undefined;
    try std.testing.expectError(Error.InvalidRSV, parseSOCKS5UDP(&[_]u8{ 0xFF, 0, 0, 0x01, 0, 0, 0, 0, 0, 0 }, &host_buf));
}

test "parseSOCKS5UDP fragment" {
    var host_buf: [64]u8 = undefined;
    try std.testing.expectError(Error.FragmentNotSupported, parseSOCKS5UDP(&[_]u8{ 0, 0, 0x01, 0x01, 0, 0, 0, 0, 0, 0 }, &host_buf));
}
